Initial implementation

tomalec opened the pull request:
### Changes proposed in this Pull Request:

This PR implements Yeoman generator:
- `ventures:github` that creates  
   ```
   CODE_OF_CONDUCT.md 
   CONTRIBUTING.md 
   ISSUE_TEMPLATE/
  ├ bug_report.md
  └ new_feature.md
   pull_request_template.md
   SECURITY.md
   ```
   The files were taken from GLA repo. Then I added a few tweaks (see [commits list](https://github.com/tomalec/generator-ventures/pull/1/commits) for more details):
      - Added emojis to names, to follow [WooCommerce example](https://github.com/woocommerce/woocommerce/issues/new/choose)
      - Added `### Additional details:` section, which was used in few other repos
      - Unified EOF line endings
      - Unified Upper case filenames
      - Added indexes to `ISSUE_TEMPLATE/*`
      - Automatically set `type: enhancement` label for `new_feature` issue template.
      - […](https://github.com/tomalec/generator-ventures/pull/1/commits) 
- `ventures:dotfiles` that creates  
   - `.editorconfig` taken from [WooCommerce trunk](https://github.com/woocommerce/woocommerce/blob/dabd2c6bf23344307a42be667f4757326235b391/.editorconfig)
   - `.gitignore` taken from GLA
   
   


### Screenshots:
![yo ventures](https://user-images.githubusercontent.com/17435/148696052-4ea2d847-af97-475b-a877-249eac79b13d.gif)



### Detailed test instructions:

1. clone this repo
2. run `npm link` from the cloned directory
3. call `yo ventures` (`yo ventures:github`) in this repo, or an other Ventures repo.



### Additional details:

- `.gitignore` will be improved in a next PR

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

tomalec committed:
Inital commit.

tomalec committed:
Add generated files.

Run `yo ventures` locally.

tomalec committed:
Add whitespace after numbered list markers in PR template.

tomalec committed:
Move expected & actual behavior sections in bug_report.md,

to state te result after the steps to reproduce and to match AutomateWoo & Pinterest for WooCommerce repos.

tomalec committed:
Document sources of dotfiles content.

tomalec committed:
Add a comment to PR template's test instructions section.

Copied from https://github.com/woocommerce/facebook-for-woocommerce/blob/3b881dae8942bdbeac871c8cbf6f418fb482e407/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L10.

tomalec committed:
Add problem and solution sections, to new_feature template.

Copied from https://github.com/woocommerce/pinterest-for-woocommerce/blame/276776d026e7c35eb5918d75a4c18fe6d5a91663/.github/ISSUE_TEMPLATE/enhancement.md#L10-L21.

tomalec committed:
Rename PR template to all caps.

To match the convention of WooCommerce and a number of our plugins.

tomalec committed:
Re-generate self.

tomalec committed:
Add indexes to `ISSUE/TEMPLATE/*`s,

to make it easier to sort them, once we get more.

tomalec committed:
Automatically set `type: enhancement` label,

for new_feature issue template.

tomalec committed:
Unify EOF line endings.

puntope commented on .gitignore:
```diff
@@ -0,0 +1,31 @@
+coverage
+
+# Directories/files that may appear in your environment
+.DS_Store
+Thumbs.db
+wp-cli.local.yml
+*.sql
+*.tar.gz
+*.tgz
+*.zip
+.idea
+.vscode/
```
💅 

In my opinion items like. `.DS_Store`, `Thumbs.db` `.idea`, `.vim` etc as well as `*.log`, `*.bak` and other files which are potentially affected by the developer environment or preference shouldn't be added to the projects' `.gitignore`.

Instead I'd suggest that each developer should be responsible of `gitignore global` as an example check here  https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

Ideally, In the project it should be only the project based ignores. Otherwise each new developer using a different environment or naming for backup or log files would pollute with more and more stuff the project `gitignore` file.

tomalec replied:
As I mentioned in OP, I'd like to improve `gitignore` template later. 
I added it here mostly, to show-case, what I plan to be there.

Initially, I was thinking of only adapting it so it works well for other repos that have different files structure, like `/assets`, etc.
But this is a really valid point. However, probably needs to be agreed with the Team, not to confuse people, but suddenly not ignore all the env files.

I created a separate issue to track that topic https://github.com/tomalec/generator-ventures/issues/3

Are you fine with leaving it as is, or should we rather remove `.gitignore` completely as YAGNI for today?

puntope replied:
Both options are fine 👌  Thanks!

puntope commented on generators/github/USAGE:
```diff
@@ -0,0 +1,3 @@
+Description:
+	Creates a bunch of `.github/` files:
+	Issue & PR templates, CODE_OF_CONDUCT, CONTRIBUTING and SECURITY.md.
```
Newline missed. 

I wonder always about why was this newline thing... seems related to the way POSIX considers a line.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

![Screenshot 2022-01-12 at 18 13 03](https://user-images.githubusercontent.com/5908855/149188690-906b14c4-25e9-4811-813a-99db8f492e8e.png)

puntope commented on generators/dotfiles/index.js:
```diff
@@ -0,0 +1,10 @@
+			this.destinationPath( './' )
+		);
+	}
+};
```
Missed newline 

See https://github.com/tomalec/generator-ventures/pull/1/files#r783281710

puntope commented on generators/dotfiles/USAGE:
```diff
@@ -0,0 +1,2 @@
+Description:
+	Creates a minimal set of dotfiles that we share across all projects.
```
Missed newline. 

See https://github.com/tomalec/generator-ventures/pull/1/files#r783281710

puntope commented on generators/app/index.js:
```diff
@@ -0,0 +1,11 @@
+	  	this.composeWith( { Generator: GitHubGenerator, path: packageRoot + '../github/index.js' } );
+	  	this.composeWith( { Generator: DotFilesGenerator, path: packageRoot + '../dotfiles/index.js' } );
+	}
+};
```
Missed newline. 

See https://github.com/tomalec/generator-ventures/pull/1/files#r783281710

puntope commented on generators/app/USAGE:
```diff
@@ -0,0 +1,7 @@
+Example:
+	ventures
+
+	This will compose ventures:github and ventures:dotfiles that could be as well used separately.
```
Missed newline. 

See https://github.com/tomalec/generator-ventures/pull/1/files#r783281710

puntope commented on generators/github/index.js:
```diff
@@ -0,0 +1,28 @@
+			{ ...this.answers }
+		);
+	}
+};
```
Missed newline. 

See https://github.com/tomalec/generator-ventures/pull/1/files#r783281710

puntope commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+## Feature Requests
+
+Feature requests can be submitted to our [ideas board](https://ideas.woocommerce.com/forums/133476-woocommerce). Be sure to include a good description of the expected behavior and use case, and before submitting a request, please search for similar ones.
```
❓ 
Are we using this for the ideas? Thought we had Canny

tomalec replied:
AFAIK, We have Canny for AutomateWoo ones only.

puntope commented on .github/ISSUE_TEMPLATE/1-bug_report.md:
```diff
@@ -0,0 +1,27 @@
+name: 🐞 Bug report
+about: Report a bug if something isn't working as expected
+title: ''
+labels: ''
```
❓ 💅 

Makes sense to add `type:bug` as label?

tomalec replied:
I was thinking about it as well. But I thought we don't have it in our new repos like GLA for a reason. For example, we want a team member/maintainer to approve something as a bug.

puntope commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,38 @@
+### Changes proposed in this Pull Request:
```
❓ 
What is the difference between [PULL_REQUEST_TEMPLATE.md](https://github.com/tomalec/generator-ventures/blob/dd68127b912804ebff894d4416df7c1dd5a979d2/.github/PULL_REQUEST_TEMPLATE.md) and this? Why we need both?

tomalec replied:
Thanks for catching. this is a leftover from before https://github.com/tomalec/generator-ventures/pull/1/commits/f0f04c1b21d447fbc23107e8ed738c9e0ff2e8fb

tomalec replied:
Fixed in https://github.com/tomalec/generator-ventures/pull/1/commits/15371b77272dbda8a504d2d7babcc24a2981a98c

puntope commented on README.md:
```diff
@@ -1 +1,28 @@
+
+## Prerequisites
+
+0. Install `npm`
+1. Install Yeoman globally `npm i -g yo`.
+2. Install this generator globally `npm i -g generator-ventures`.
+	That will be available once we publish this package.
+	Before that, you can:
```
💅 

I'd add this after publishing it

tomalec replied:
Do you mean to remove l10-l12? And leave only currently working steps from l13-14

puntope replied:
> Do you mean to remove l10-l12

Yes, sorry I added it to more lines by mistake. 

But in my case I was doing the steps 0., .1, .2, and then in step 2 I realised after my failed attempt the rest of the instructions.

puntope replied:
https://github.com/yeoman/yo/blob/main/lib/usage.txt 

In here they used this format. But maybe is a completely different thing

tomalec replied:
WDYT of 
```md
2. Install this generator globally
	- Before the package is published you need to clone and link it manually:
		1. clone this repo
		2. run `npm link` from the cloned directory
	- Once the package is published `npm i -g generator-ventures` would work.
```
I'd like to keep the `npm i` way as if we'd eventually publish it, we need the README instruction for the published package be in the package before we hit `npm publish` ;) So, I thought let's add them today, not to forget them in the future.

puntope replied:
👌  Nice, thanks for changing it.

puntope commented on generators/app/USAGE:
```diff
@@ -0,0 +1,7 @@
+Description:
```
❓  Is this specific of Yeoman to not have file extension?

tomalec replied:
Yes, I'm not sure if it's only Yeomans conventions, but that how it looked in examples, and that's how it works when you do 
`yo ventures --help`

puntope replied:
<img width="1040" alt="Screenshot 2022-01-12 at 21 08 08" src="https://user-images.githubusercontent.com/5908855/149214064-c51d7a1b-91c7-4473-b0d6-49c7f63c2a15.png">

This is what the console renders for me as the help. However doesn't really makes sense for me. Especially

Usage:
`yo ventures:app [options]` I guess app is `dotfiles|github` but is not clear, since it's team tool I would make it clearer if possible. 

Example:
Should be `yo ventures` right? In the help is written just `ventures`

tomalec replied:
`yo ventures:app [options]` is generated by Yeoman, as there is a convention that `generatorfoo:app` is the default = `generatorfoo`. I'd take a look if we could change that, but could be tricky.

> I guess app is `dotfiles|github`
app could be nothing == `app`, `dotfiles`, or `github`

You can check the `yo ventures:github --help` as well.

> Example:
> Should be yo ventures right? In the help is written just ventures

Yes, I'm going to change that.

tomalec replied:
WDYT of 
```
Usage:
  yo ventures[:app] [options]
```
Added in https://github.com/tomalec/generator-ventures/pull/1/commits/df45a7b94769cb26068af304a273048718347d93

puntope replied:
That looks clearer (for me at least). Thanks for changing it

puntope commented on generators/github/index.js:
```diff
@@ -0,0 +1,28 @@
+				type: "input",
+				name: "title",
+				message: "Your project title (used in CONTRIBUTING.md)",
+				default: this.packageJson.get( 'title' ) || this.appname
```
❓ 
Where we defined this `this.appname` variable ?

tomalec replied:
This is Yeoman thing https://yeoman.github.io/generator/Generator.html

puntope replied:
Interesting. Is getting name of the directory, ok

tomalec committed:
Add missing EOF line endings.

Addresses https://github.com/tomalec/generator-ventures/pull/1#pullrequestreview-850772432.

tomalec commented:
Thanks for the review :) I addressed newlines, and created a separate issue for `.gitignore`. I think it's ready for another round :)

puntope commented:
> Thanks for the review :) I addressed newlines, and created a separate issue for `.gitignore`. I think it's ready for another round :)

Yes, there are a few ❓ as pending still. 

Besides that, one thing I just saw right now is that this repo is outside ventures. Is that for some reason?

tomalec commented:
> Besides that, one thing I just saw right now is that this repo is outside ventures. Is that for some reason?

I don't have privileges to create a repo in `/woocommerce/` org.

tomalec committed:
Remove leftover `.github/pull_request_template.md`

Addresses https://github.com/tomalec/generator-ventures/pull/1#discussion_r783289612

puntope reviewed (approved):
✅  LGTM

- I could generate .github and .dotfiles easily ✅ 
- I could use `yo ventures`, `yo ventures:github` and `yo ventures:dotfiles` successfully 

💅  pending some minor discussion around documentation and `.gitignore` to be addressed and discussed in future iterations if desired

tomalec committed:
Fix examples in USAGE

Addresses https://github.com/tomalec/generator-ventures/pull/1#discussion_r783414643

tomalec committed:
Update default app usage,

to show that `:app` is optional.

Addresses https://github.com/tomalec/generator-ventures/pull/1#discussion_r783411371

tomalec committed:
Reword README, to emphasize the current need of local cloning.

Addresses https://github.com/tomalec/generator-ventures/pull/1#discussion_r783291225.
